pkg/archive/archive.go

```diff
@@ -660,7 +660,17 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
 		mask = *forceMask
 	}
 
-	switch hdr.Typeflag {
+	typeFlag := hdr.Typeflag
+
+	if forceMask != nil {
+		// If we have a forceMask, force the real type to either be a directory,
+		// or a regular file.
+		if typeFlag != tar.TypeDir {
```

Comment:
What about symlinks? This turns them into regular files, and the destination seems to be lost.